libraries/Microsoft.Bot.Builder.LanguageGeneration/LGFile.cs

```diff
@@ -0,0 +1,350 @@
+using System.Linq;
+using System.Text;
+using Microsoft.Bot.Expressions;
+using Microsoft.Bot.Expressions.Memory;
+
+namespace Microsoft.Bot.Builder.LanguageGeneration
+{
+    /// <summary>
+    /// LG entrance, including properties that LG file has, and evaluate functions.
+    /// </summary>
+    public class LGFile
+    {
+        private readonly ExpressionEngine expressionEngine;
+
+        public LGFile(
+            IList<LGTemplate> templates = null,
+            IList<LGImport> imports = null,
+            IList<Diagnostic> diagnostics = null,
+            IList<LGFile> references = null,
+            string content = null,
+            string id = null,
+            ExpressionEngine expressionEngine = null,
+            ImportResolverDelegate importResolver = null)
+        {
+            Templates = templates ?? new List<LGTemplate>();
+            Imports = imports ?? new List<LGImport>();
+            Diagnostics = diagnostics ?? new List<Diagnostic>();
+            References = references ?? new List<LGFile>();
+            Content = content ?? string.Empty;
+            Id = id ?? string.Empty;
+            this.expressionEngine = expressionEngine ?? new ExpressionEngine();
+            this.ImportResolver = importResolver;
+        }
+
+        public IList<LGTemplate> AllTemplates
+        {
+            get
+            {
+                var referenceTemplates = References.GroupBy(x => x.Id).Select(x => x.First()).SelectMany(x => x.Templates);
+                return Templates.Union(referenceTemplates).ToList();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets delegate for resolving resource id of imported lg file.
+        /// </summary>
+        /// <value>
+        /// Delegate for resolving resource id of imported lg file.
+        /// </value>
+        public ImportResolverDelegate ImportResolver { get; set; }
+
+        /// <summary>
+        /// Gets or sets templates that this LG file contains directly.
+        /// </summary>
+        /// <value>
+        /// templates that this LG file contains directly.
+        /// </value>
+        public IList<LGTemplate> Templates { get; set; }
+
+        /// <summary>
+        /// Gets or sets import elements that this LG file contains directly.
+        /// </summary>
+        /// <value>
+        /// import elements that this LG file contains directly.
+        /// </value>
+        public IList<LGImport> Imports { get; set; }
+
+        /// <summary>
+        /// Gets or sets all references that this LG file has from <see cref="Imports"/>.
+        /// </summary>
+        /// <value>
+        /// import elements that this LG file contains directly.
+        /// </value>
+        public IList<LGFile> References { get; set; }
+
+        /// <summary>
+        /// Gets or sets diagnostics.
+        /// </summary>
+        /// <value>
+        /// diagnostics.
+        /// </value>
+        public IList<Diagnostic> Diagnostics { get; set; }
```

Comment:
Let's make sure this diagnostics only contains the diagnostics belonged to this LGFile, and similar to `AllTemplates`, let's create AllDialognotics to capture all diagnostics from all references.